### PR TITLE
Persistence

### PR DESCRIPTION
**Pull request by hpx7**

### CODE REVIEW

File: templates/base/server/.rtag/logstore.ts.hbs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    const rows: { time: number; record: Buffer }[] = [];
+    while (fs.readSync(fd, metadataBuf, 0, metadataBuf.length, null) > 0) {
+      const time = metadataBuf.readBigUInt64LE();
+      const len = metadataBuf.readUInt16LE(8);
+      const recordBuf = Buffer.alloc(len);
+      fs.readSync(fd, recordBuf, 0, recordBuf.length, null);
+      rows.push({ time: Number(time), record: recordBuf });
+    }
+    return rows;
```
Comment (mihirpandya) [inline templates/base/server/.rtag/logstore.ts.hbs:39]: Close file? https://nodejs.org/api/fs.html#fs_dir_closesync

File: templates/base/server/.rtag/logstore.ts.hbs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  public append(partitionId: string, time: number, record: Uint8Array) {
+    if (!this.partitionFiles.has(partitionId)) {
+      this.partitionFiles.set(partitionId, fs.openSync(path.join(this.storageDir, partitionId), "a"));
+    }
+    const fd = this.partitionFiles.get(partitionId)!;
+
+    const buf = Buffer.alloc(12);
+    buf.writeBigUInt64LE(BigInt(time));
+    buf.writeUInt16LE(record.byteLength, 8);
+    fs.appendFileSync(fd, buf);
```
Comment (mihirpandya) [inline templates/base/server/.rtag/logstore.ts.hbs]: Why not append both `buf` and `record` together in one synchronous call? If the server is stopped after appending `buf` but not `record`, you kinda have an invalid/corrupt record.

File: templates/base/server/.rtag/logstore.ts.hbs [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import fs from "fs";
+import path from "path";
+
+export default class LogStore {
+  private storageDir: string;
+  private partitionFiles: Map<string, number>;
```
Comment (mihirpandya) [inline templates/base/server/.rtag/logstore.ts.hbs:6]: what happens when the server is bounced? will these file descriptors be closed gracefully?

Comment (hpx7) [reply templates/base/server/.rtag/logstore.ts.hbs:6]: yeah the os will reclaim file descriptors on process exit

File: templates/base/server/.rtag/logstore.ts.hbs (hunk 1 above)
Comment (hpx7) [reply templates/base/server/.rtag/logstore.ts.hbs:39]: good call, fixed

File: templates/base/server/.rtag/logstore.ts.hbs (hunk 2 above)
Comment (hpx7) [reply templates/base/server/.rtag/logstore.ts.hbs]: done

**Review by mihirpandya — APPROVED** [review]

**Pull request merged**